Filters for integration page

kknopper opened the pull request:
## Proposed change
A rework of #24642 / #24500 / #18512 to add proper filtration to the integrations page. Also addresses https://github.com/home-assistant/home-assistant.io/issues/9944

[Netlify preview / demo](https://deploy-preview-33644--home-assistant-docs.netlify.app/integrations/?brands=featured)

Saw the new site changes, and finally came back to rework this PR. Merged in latest site changes, and updated code to remove jquery element selectors. Also added `Brand Partners` to category dropdown

used @ludeeus's notes from here as a guideline [#16996 (comment)](https://github.com/home-assistant/home-assistant.io/issues/16996#issuecomment-880515679)

-  Better styling for filters. Thanks @ludeeus for the starting point
-  added filtration of IoT Classes, and Quality Scale
- all filtration items have been moved into `<select>` dropdowns. (Category, Version, IoT, and Quality)
-  It's now possible to combine filters (Search, IoT, Version, Category, and Quality)
-  an easy way to reset filters (buttons get added to the dom that can be clicked to remove filters)
-  complete rework of the `applyFilter()` function that allows for combining filters
-  update is backwards compatible with old `#` urls
-  on page load, if no filters are present in the url, the category is set to `Featured` like the production site
-  integrations can now be searched for by their iot class and quality scale. You can also search for Featured integrations

Completely reworked the javascript for the integrations page.
I opted to use the [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) browser API, over the hash method, as this api makes it easy to add and remove querys/filters to the url.

Created a `buildQueryFromURL()` function to build a query object of active filters any time the url gets updated. This object is then compared against the integrations array for filtration. Added more helper functions as well.


### Functionality:
- Old # Urls still work as expected. Also allowed for new hash urls for IoT and Quality which look like so:
`.../integrations/#iot_class/local-polling` and `.../integrations/#quality_scale/internal`

- If a `#` url is present, any change to a filter will convert the URL to use the URLSearchParams API format instead.

- On page load, or any change to the url, the current active filters (select dropdowns / search bar) will get updated to match the current active filters

- Filter removal buttons are dynamically added to the dom as well, and also update to match current active filters

- Filters can be removed by clicking these buttons, or setting a filter dropdown to its initial state

### Caveats:
- URLSearchParams is not supported by IE
- this update contains a lot of ES6 javascript which is not supported by IE
- if IE needs to be supported, there are [polyfills](https://github.com/jerrybendy/url-search-params-polyfill) available for the - URLSearchParams api

This is a pretty large change, because of this, I tried to make my code as readable as possible with code comments for new functions.


## Type of change
<!--
    What types of changes does your PR introduce to our documentation/website?
    Put an `x` in the boxes that apply. You can also fill these out after
    creating the PR.
-->

- [x] Spelling, grammar or other readability improvements (`current` branch).
- [ ] Adjusted missing or incorrect information in the current documentation (`current` branch).
- [ ] Added documentation for a new integration I'm adding to Home Assistant (`next` branch).
  - [ ] I've opened up a PR to add logos and icons in [Brands repository](https://github.com/home-assistant/brands).
- [ ] Added documentation for a new feature I'm adding to Home Assistant (`next` branch).
- [ ] Removed stale or deprecated documentation.

## Additional information
<!--
    Details are important, and help maintainers processing your PR.
    Please be sure to fill out additional details, if applicable.
-->

- Link to parent pull request in the codebase: 
- Link to parent pull request in the Brands repository: 
- This PR fixes or closes issue: fixes #9944 

## Checklist

- [x] This PR uses the correct branch, based on one of the following:
  - I made a change to the existing documentation and used the `current` branch.
  - I made a change that is related to an upcoming version of Home Assistant and used the `next` branch.
- [x] The documentation follows the Home Assistant documentation [standards].

[standards]: https://developers.home-assistant.io/docs/documenting/standards


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Enhanced component search with a minimum height for better layout.
  - Improved integration filtering with new options based on categories, versions, IoT classes, and quality scales.
  - Added dynamic filtering and search functionality.
  - Introduced radio buttons and dropdowns for refined filter options.
  - Dynamic URL updates to reflect active filters.

- **Style**
  - Expanded styling capabilities with new classes for layout adjustments, filter buttons, and alert messages.
  - Updated styles for various components including search inputs, filters, and alert messages to improve the visual appearance and usability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
overhaul of filters on integrations page

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
addressing PR review feedback

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
min-height fix for filter button ui

Commit:
removing console logs

Commit:
removing counts from select dropdowns

Commit:
Other dropdown select fix

Commit:
removing hyphens from filter buttons

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
capitalizing quality scale to match other dropdowns

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
adding check for dropdowns, better styling for dropdowns, adjusting all selector

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
fixing select dropdown color on mobile

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
fixing a bug with the filter removal buttons

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
adjusted search input text so that it stays consistent with url and url pushState

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
merging with latest site updates

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
removing gemfile change

frenck commented:
@kknopper This is a good time to add this now!

I do think we need to fix one "big" issue, though, which is a path most users will walk:


https://github.com/home-assistant/home-assistant.io/assets/195327/e90d8b46-909e-45c7-915d-03abfb7eef0f

What happens in this case is you go to integration and search for what you are looking for, but it ain't clear for a regular visitor. Its search is limited to the featured filter.

While we do want to show featured when entering the page, the search should search everything in this case (and thus ignore the featured filter).

frenck commented:
Last but not least, the "Partner brands" are now kinda hidden; which makes me wonder:

Should featured & partner brands be "hidden" in categories at all? Of should they should up as single click buttons?

kknopper commented:
for the search bar - you're right, right now it works more as an additional "filter" than a standalone search. Would you like it to revert to the "all" filters only on the first use?

Also for the Featured / Partner brands - i can try making those different filters, separate from the category dropdown

frenck commented:
> for the search bar - you're right, right now it works more as an additional "filter" than a standalone search. Would you like it to revert to the "all" filters only on the first use?

Yeah I think that works. If the page is visited, without any filters or search terms, automatically only show featured ones (without marking the "featured" category), and just clear that to all once anything is used (search, other filters, and what not).

> Also for the Featured / Partner brands - i can try making those different filters, separate from the category dropdown

Yeah sounds good. My main concern is partners, they have invested in our community to offer devices that are tested and work together with the project to ensure stuff works. We should reward them for that and make sure they stand out in a way (and not just hidden in a category menu).

../Frenck

Commit:
adding brands filter, code cleanup

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper commented:
- Added a new `Brands` filter. I added 3 radio buttons above the select dropdowns, these buttons toggle the new filter, and `All` removes the filter. I opted to make these radio buttons so that the filter functionality would be similar to the select dropdowns (only 1 option can be selected at a time).

- The following urls `.../integrations/#featured`, `.../integrations/#works-with-home-assistant` now will toggle the radio buttons and work as expected. Changing the filter in the UI will update the url to the newer query string version.

- For the search bar, it is set to clear all filters on the first search, unless a query filter is detected in the url on page load.

- Also updated the `.../intregrations/#all` url so that it clears all filters (sets all filters to All)

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
better styling for radio buttons

Commit:
make sure search input is always lowercase

kknopper commented:
This PR is ready for another review

SeanPM5 commented:
I like this, makes it super easy to find local controlled integrations. A few small suggestions:

* I think "Quality Scale" select options should be sorted by quality rather than alphabetically. Platinum > Gold > Silver etc. Most using this filter are trying to find the highest quality integrations.
* Like above, I think "IoT Class" select options should also be sorted from best to worst instead of alphabetically. Local Push > Local Polling > Cloud Push > Cloud Polling > Assumed State.
* IMO, initially clicking on the "Featured" and "Partners" radio buttons should clear all existing filters so that users always see all 18 featured integrations and all 8 partner integrations.
* For the no results message, instead of it saying "Nothing found!" I think the wording could be tweaked to something a bit more informative like "No results matching this filter" or something along those lines.

Commit:
fixed merge conflict

Commit:
radio buttons now clear filters, cleaned up codebase, renamed functions, broke complex functions into smaller functions

Commit:
updated iot and quality dropdown order to be manual

kknopper commented:
Ready for another review, feedback addressed:
- adjusted quality / iot dropdowns order
	- Note: this order had to be done manually (couldnt find a way to  do a custom sort in liquid) - this means if filters are added in the future, they will need to be added manually to this page
-  updated `Nothing Found` message to `No results matching this filter` and centered text
- updated functionality of radio buttons so that if you click `Featured` or `Partners` all other active filters get cleared
	- Note: right now, the filters dont reset for `All`, it will just removed the `Brands` filter


- Cleaned up codebase to make it easier to set custom filters in the code if needed in the future
	- (Previously it was setup to add / remove filters, but not completely overwrite filters with a new custom one) 
	- Renamed some functions to make functionality more clear
	- Broke down some complex functions into smaller functions to make code more readable
	- updated some comments to better explain code functions

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper commented:
@frenck Please let me know if there is any additional feedback, or design changes you'd like to see in this PR :)

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

frenck committed:
Merge branch 'current' into integrations-filter-overhaul

kisst commented:
Why is this isn't merged yet ? Anything pending or just needs the reviews ?
( my attempt to get some attention , and making sure it doesn't go stale again )

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

Commit:
Updating Quality Scale filter to match 2024.12 release

Commit:
variable name fix

kknopper commented:
Ready for another review!

Updated Quality Scale filter to match latest updates from the 2024.12 release (added bronze, and legacy options). Also added emoji icons to match the docs page. https://www.home-assistant.io/docs/quality_scale/

Note: did not add `custom` option as no integrations official integrations get those labels, and did not add `no scale` since those integrations are simply not labeled with a scale

Observation: I also noticed some integrations that were previously labeled Platinum no longer seem to have that label. Not sure if all integrations are being re-evaluated to match the latest criteria? (WLED, Hue for example)

Commit:
fixed issue where search bar was not being clearing as expected when selecting featured or brands radio buttons

Commit:
cleanup, removing console logs

kknopper commented:
Fixed a small bug where the search input was not being cleared when selecting `Featured` or `Brands` - clicking these remove all other filers, and should clear the search bar. This has been fixed.

I think this PR is in a pretty good place now and I have not been able to find any other related issues with expected behavior

ChadNedzlek commented:
Please do not close this. @kknopper can you consider merging this?

Commit:
css font cleanup, search input cleanup

kknopper commented:
- Cleaned up css for emoji rendering in Safari for select dropdown and active filter buttons.
- Merged in logic from #36844 to clean up search input.
  - `lastSearchText` prevents a new filter from applying if the search value doesnt change after debounce
  - also changed search input event listener from `keyup` to `input` to be safer. (text has to be entered to fire)

kknopper commented:
@ChadNedzlek im just a contributor, a HA team member needs to help give a final review of my work

kknopper committed:
Merge branch 'current' into integrations-filter-overhaul

ChadNedzlek commented:
Indeed. Just noticed it had a bunch of merge conflicts... how does one get the attention of a team member? Just wait and hope?

ChadNedzlek commented:
(Also I needed to do _something_ so the annoying bot didn't auto-close it)

silamon commented on source/integrations/index.html:
```diff
@@ -23,38 +23,37 @@
+          <h3 class="filter-title">Category</h3>
+          <select class="ha_category" name="category" data-id="cat">
+            <option value="#">All</option>
+             {%- for category in categories -%}
```
Can you fix the formatting here? The assigns are within the for and break out.

silamon commented on source/integrations/index.html:
```diff
@@ -310,25 +765,28 @@
-          : '#all';
-        history.pushState('', '', newHash);
+    searchInputEl.addEventListener('input', debounce(() => {
+      let value = searchInputEl.value.toLowerCase();
+      // sanitize input
+      value = value.replace(/[(\?|\&\{\}\(\))]/gi, '').trim();
```
```suggestion
      let value = searchInputEl.value
         .toLowerCase()
         // sanitize input
         .replace(/[(\?|\&\{\}\(\))]/gi, '')
         .trim();
```

silamon commented on source/integrations/index.html:
```diff
@@ -310,25 +765,28 @@
+          searchInputDirty = true;
+        }
+
+        if (value === lastSearchText) return;
+
+        lastSearchText = value;
+
+        setQueryURL('search', value);
+        applyFilter();
```
```suggestion
        // Only apply filter if value has changed
        if (value !== lastSearchText) {
          lastSearchText = value;
          setQueryURL('search', value);
          applyFilter();
        }
```

silamon commented on source/integrations/index.html:
```diff
@@ -264,6 +418,296 @@
+        }
+    }
+
+    // remove buttons that don't match current filter // used in updateFilterButtons
+    function removeFilterButtons(buttons, query) {
+      let buttonsToRemove = buttons.filter(button => {
+          for (key in query) {
+            if (button.dataset.id === key) {
+              return false;
+            }
+          }
+          return true;
+        });
+        buttonsToRemove.forEach(button => button.remove());
+    }
```
```suggestion
    // remove buttons that don't match current filter // used in updateFilterButtons
    function removeFilterButtons(buttons, query) {
      let buttonsToRemove = buttons.filter(button => {
        for (key in query) {
          if (button.dataset.id === key) {
            return false;
          }
        }
        return true;
      });
      buttonsToRemove.forEach(button => button.remove());
    }
```

silamon committed:
Formatting issues all done

silamon committed:
Reformatting again

silamon committed:
More formatting

silamon committed:
Update source/integrations/index.html

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

silamon committed:
Update source/integrations/index.html

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

silamon commented on source/integrations/index.html:
```diff
@@ -23,68 +23,38 @@
 {%- assign components_by_version = site.integrations | group_components_by_release -%}
 {%- assign categories = components | map: 'ha_category' | join: ',' | downcase | split: ',' | uniq | sort -%}
+{%- assign iot_classes = "Local Push,Local Polling,Cloud Push,Cloud Polling,Assumed State,Calculated,Configurable" |
+join: ',' | join: ',' | split: ',' -%}
```
```suggestion
join: ',' | split: ',' -%}
```

silamon committed:
Update source/integrations/index.html

silamon reviewed (dismissed):
Congratulations, and thanks for having patience! I've done a few formatting things and small nitpicks and if everything still works, I'm going to merge it in.

silamon committed:
Update index.html
